sans/direct-beam.ipynb

```diff
@@ -0,0 +1,529 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "bad_straws = sc.Variable(dims=['spectrum'], shape=sample.coords['spectrum'].shape, dtype=sc.dtype.bool)\n",
+    "mask = sc.Variable(value=True)\n",
+    "bad_straws[loki.straw(tube=0, straw=6)] |= mask\n",
+    "bad_straws[loki.straw(tube=1, straw=2)] |= mask\n",
+    "bad_straws[loki.straw(tube=5, straw=1)] |= mask\n",
+    "for straw in 1,2,3,4,5:\n",
+    "    bad_straws[loki.straw(tube=12, straw=straw)] |= mask\n",
+    "masks['bad-straws'] = bad_straws"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Bad pixels"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "scrolled": false
+   },
+   "outputs": [],
+   "source": [
+    "plot(pixel_counts, vmax=100, axes=['straw', 'tube', 'pixel'])"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "pos = sc.neutron.position(sample)\n",
+    "x = sc.geometry.x(pos)\n",
+    "y = sc.geometry.y(pos)\n",
+    "counts = spectrum_counts.data\n",
+    "counts.variances = None # TODO support comparison with variances\n",
+    "masks['electronics-error'] = sc.less(sc.abs(x), 0.2 * sc.units.m) \\\n",
+    "                             & sc.less(sc.abs(y), 0.03 * sc.units.m) \\\n",
+    "                             & sc.equal(counts, 0.0 * sc.units.counts)\n",
+    "print(f\"Masking {sc.sum(masks['electronics-error'], 'spectrum').value} bad pixels due to electronics error.\")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "tof = sample.coords['tof']\n",
+    "masks['promt-pulse'] = sc.less(tof['tof',1:], 1500.0 * sc.units.us) | \\\n",
+    "                      (sc.greater(tof['tof',:-1], 17500.0 * sc.units.us) & \\\n",
+    "                       sc.less(tof['tof',1:], 19000.0 * sc.units.us))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Note that this needs more tuning and masks too much. Better do this after moving detectors?\n",
+    "masks['beam-stop'] = sc.less(sc.abs(x), 0.03 * sc.units.m) \\\n",
+    "                   & sc.less(y, 0.028 * sc.units.m) \\\n",
+    "                   & sc.greater(y, -0.016 * sc.units.m)\n",
+    "pixel_counts.masks['beam-stop'] = sc.reshape(masks['beam-stop'], dims=pixel_counts.dims, shape=tuple(pixel_counts.shape))\n",
+    "plot(pixel_counts, vmax=1000, axes=['straw', 'tube', 'pixel'])"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "masks['tube-ends'] = sc.greater(x, 0.36 * sc.units.m) \\\n",
+    "                   | sc.less(x, -0.36 * sc.units.m)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### X-Y projection\n",
```

Comment:
Not clear to me why this is required.

Reply:
It is not (for the purpose of this script), just showing off. :)

It is "required" as a visualization of raw that that is more useful to instrument scientists and users: With 3D / volumetric detectors the instrument view has limited use for getting a quick preview of the data (it is probably more of a debugging / commissioning tool?). This is something that has been requested in the past.